start.sh

```diff
@@ -0,0 +1,147 @@
+###############################################################################
+
+if [ -f /etc/os-release ]; then
+  source /etc/os-release
+  DISTRO_ID=$ID
+elif [ -f /etc/issue ]; then
+  # Fallback for older systems
+  DISTRO_ID=$(head -n1 /etc/issue | awk '{print tolower($1)}')
+else
+  DISTRO_ID="unknown"
+fi
+
+###############################################################################
+# Check if Docker is installed
+###############################################################################
+
+function install_docker_ubuntu() {
+  echo "[INFO] Installing Docker for Ubuntu..."
+  sudo apt-get update && sudo apt-get install -y docker.io
+  sudo systemctl enable docker
+  sudo systemctl start docker
+}
+
+function install_docker_arch() {
+  echo "[INFO] Installing Docker for Arch..."
+  sudo pacman -Sy --noconfirm docker
+  sudo systemctl enable docker
+  sudo systemctl start docker
+}
+
+if ! command -v docker &> /dev/null; then
+  echo "[INFO] Docker not found on your system."
+
+  if [[ "$DISTRO_ID" == "ubuntu" ]]; then
+    install_docker_ubuntu
+  elif [[ "$DISTRO_ID" == "arch" ]]; then
+    install_docker_arch
+  else
+    echo "[WARNING] Your distro is '$DISTRO_ID', which isn't supported by this script."
+    echo "          Please follow Docker's official instructions at:"
+    echo "          https://docs.docker.com/get-docker/"
+    exit 1
+  fi
+
+  # Add current user to docker group
+  echo "[INFO] Adding '$USER' to the 'docker' group..."
+  sudo groupadd docker 2> /dev/null || true
+  sudo usermod -aG docker "$USER"
+
+  echo "[INFO] Docker installation complete."
+  echo "      You may need to log out and log back in (or restart your shell) for the group"
+  echo "      membership to take effect. Please do that, then re-run this script."
+  exit 0
+else
+  echo "[OK] Docker is already installed."
+fi
+
+# Double-check Docker is running properly
+if ! docker ps &> /dev/null; then
+  echo "[ERROR] Docker is installed but not running. Please start Docker (e.g., 'sudo systemctl start docker')."
+  exit 1
+fi
+
+###############################################################################
+# Check if the Docker image already exists
+###############################################################################
+# assume that Dockerfile is in the same directory
+echo "[INFO] Checking if the Docker image '$IMAGE_NAME' exists..."
+
+IMAGE_EXISTS=$(docker images -q "$IMAGE_NAME")
+
+if [ -z "$IMAGE_EXISTS" ]; then
+    echo "[INFO] '$IMAGE_NAME' not found. Building image from Dockerfile..."
+    
+    # Build the image
+    docker build -t "$IMAGE_NAME" "$DOCKERFILE_DIR"
+else
+    echo "[OK] Docker image '$IMAGE_NAME' already exists."
+fi
+
+###############################################################################
+# dockerfile hash
+###############################################################################
+# 1. Compute current hash of Dockerfile
+CURRENT_HASH="$(md5sum "$DOCKERFILE_PATH" | awk '{print $1}')"
```

Comment:
I'm pretty sure `docker build` automatically detects cached / stored layers that have previously been built and reuses them. Not sure if this section is necessary. Unless there's a different reason I'm missing?

Reply:
I was getting issues with docker not rebuilding even if I changed the dockerfile, I'll keep it in case other people face similar issues
The VS Code extension does a similar thing

Reply:
Weird. Makes sense tho.